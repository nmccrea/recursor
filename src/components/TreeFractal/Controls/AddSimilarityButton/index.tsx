import React from "react"
import { useDispatch } from "react-redux"
import { addOne } from "state/similarities/actions"
import { SimilarityInputs } from "state/similarities/types"
import { ButtonPrimary } from "components/parts/Button"

/**
 * Generate a set of new similarity inputs.
 *
 * @returns A new `SimilarityInputs` object that may be added to the store.
 */
const generateNewSimilarityInputs = (): SimilarityInputs => ({
  translation: 1,
  scale: 0.75,
  angle: 0,
  depth: 5,
  color: "black",
})

const AddSimilarityButton = () => {
  const dispatch = useDispatch()
  return (
    <ButtonPrimary
      onClick={() => dispatch(addOne(generateNewSimilarityInputs()))}
    >
      Add
    </ButtonPrimary>
  )
}

export default AddSimilarityButton
