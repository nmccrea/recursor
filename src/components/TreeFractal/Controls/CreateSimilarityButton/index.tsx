import React from "react"
import { useDispatch } from "react-redux"
import { createOne } from "state/similarities/actions"
import { ButtonPrimary } from "components/parts/Button"

const CreateSimilarityButton = () => {
  const dispatch = useDispatch()
  return (
    <ButtonPrimary onClick={() => dispatch(createOne())}>Add</ButtonPrimary>
  )
}

export default CreateSimilarityButton
