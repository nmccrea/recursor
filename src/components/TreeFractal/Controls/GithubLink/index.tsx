import React from "react"
import { useStaticQuery, graphql } from "gatsby"
import { GithubImage } from "./styled"

const GithubLink = () => {
  const { site } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            githubUrl
          }
        }
      }
    `
  )

  return (
    <a href={site.siteMetadata.githubUrl} target="blank">
      <GithubImage />
    </a>
  )
}

export default GithubLink
