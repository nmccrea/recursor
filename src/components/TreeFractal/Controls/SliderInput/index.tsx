import React from "react"
import { useSelector, useDispatch } from "react-redux"
import { BranchPatternSelectorCreator } from "../../../../state/branchPatterns/selectors"
import { BranchPatternInputActionCreator } from "../../../../state/branchPatterns/actions"
import {
  BranchPatternId,
  BranchPatternNumericInputs,
  BranchPatternNumericInputKey,
} from "../../../../state/branchPatterns/types"

interface SliderInputProps<Key extends BranchPatternNumericInputKey> {
  branchPatternId: BranchPatternId
  selectorCreator: BranchPatternSelectorCreator<BranchPatternNumericInputs[Key]>
  actionCreator: BranchPatternInputActionCreator<Key>
  label: string
  min: number
  max: number
  step: number
}

const SliderInput = <Key extends BranchPatternNumericInputKey>({
  branchPatternId,
  selectorCreator,
  actionCreator,
  label,
  min,
  max,
  step,
}: SliderInputProps<Key>) => {
  const value = useSelector(selectorCreator(branchPatternId))
  const dispatch = useDispatch()
  const domId = `${branchPatternId}/${label.toLowerCase()}`
  return (
    <>
      <label htmlFor={domId}>{label}</label>
      <input
        type="range"
        value={value}
        id={domId}
        min={min}
        max={max}
        step={step}
        onChange={(event) =>
          dispatch(actionCreator(branchPatternId, Number(event.target.value)))
        }
      />
    </>
  )
}

export default SliderInput
