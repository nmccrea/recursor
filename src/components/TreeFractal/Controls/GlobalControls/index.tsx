import React from "react"
import AddSimilarityButton from "./AddSimilarityButton"

const GlobalControls = () => (
  <div>
    <AddSimilarityButton />
  </div>
)

export default GlobalControls
