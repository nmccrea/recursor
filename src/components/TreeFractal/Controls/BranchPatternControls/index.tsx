import React from "react"
import {
  getTranslationSetterFor,
  getScaleSetterFor,
  getAngleSetterFor,
  getDepthSetterFor,
} from "../../../../state/branchPatterns/actions"
import {
  getTranslationSelectorFor,
  getScaleSelectorFor,
  getAngleSelectorFor,
  getDepthSelectorFor,
} from "../../../../state/branchPatterns/selectors"
import { BranchPatternId } from "../../../../state/branchPatterns/types"
import RemoveBranchPatternButton from "./RemoveBranchPatternButton"
import SliderInput from "./SliderInput"

interface BranchPatternControlsProps {
  branchPatternId: BranchPatternId
}

const BranchPatternControls = ({
  branchPatternId,
}: BranchPatternControlsProps) => (
  <div style={{ display: "flex" }}>
    <p>{branchPatternId}</p>

    <RemoveBranchPatternButton branchPatternId={branchPatternId} />

    <SliderInput
      selector={getTranslationSelectorFor(branchPatternId)}
      actionCreator={getTranslationSetterFor(branchPatternId)}
      label="Translation"
      min={0}
      max={1.5}
      step={0.05}
    />

    <SliderInput
      selector={getScaleSelectorFor(branchPatternId)}
      actionCreator={getScaleSetterFor(branchPatternId)}
      label="Scale"
      min={0}
      max={2}
      step={0.05}
    />

    <SliderInput
      selector={getAngleSelectorFor(branchPatternId)}
      actionCreator={getAngleSetterFor(branchPatternId)}
      label="Angle"
      // `min` and `max` must both be multiples of `step`
      min={-Math.round((2 * Math.PI) / 0.0001) * 0.0001}
      max={Math.round((2 * Math.PI) / 0.0001) * 0.0001}
      step={0.0001}
    />

    <SliderInput
      selector={getDepthSelectorFor(branchPatternId)}
      actionCreator={getDepthSetterFor(branchPatternId)}
      label="Depth"
      min={0}
      max={8}
      step={1}
    />
  </div>
)

export default BranchPatternControls
