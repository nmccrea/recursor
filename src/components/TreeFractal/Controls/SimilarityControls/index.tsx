import React from "react"
import styled from "styled-components"
import { formatNumber, formatAngle } from "../../../../utils/valueFormatter"
import {
  getTranslationSetterFor,
  getScaleSetterFor,
  getAngleSetterFor,
  getDepthSetterFor,
} from "../../../../state/similarities/actions"
import {
  getTranslationSelectorFor,
  getScaleSelectorFor,
  getAngleSelectorFor,
  getDepthSelectorFor,
} from "../../../../state/similarities/selectors"
import { SimilarityId } from "../../../../state/similarities/types"
import RemoveSimilarityButton from "./RemoveSimilarityButton"
import SliderInput from "./SliderInput"

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  padding-top: 10px;
`

const Inputs = styled.div`
  flex-grow: 1;

  display: flex;
  flex-wrap: wrap;
  align-items: baseline;
  justify-content: space-evenly;
`

interface SimilarityControlsProps {
  similarityId: SimilarityId
}

const SimilarityControls = ({ similarityId }: SimilarityControlsProps) => (
  <Container>
    <div>
      <RemoveSimilarityButton similarityId={similarityId} />
    </div>

    <Inputs>
      <SliderInput
        label="Angle"
        // `min` and `max` must both be multiples of `step`
        min={-Math.round((2 * Math.PI) / 0.0001) * 0.0001}
        max={Math.round((2 * Math.PI) / 0.0001) * 0.0001}
        step={0.0001}
        selector={getAngleSelectorFor(similarityId)}
        actionCreator={getAngleSetterFor(similarityId)}
        valueFormatter={formatAngle}
      />

      <SliderInput
        label="Scale"
        min={0}
        max={2}
        step={0.05}
        selector={getScaleSelectorFor(similarityId)}
        actionCreator={getScaleSetterFor(similarityId)}
        valueFormatter={(value) => formatNumber(value, { precision: 2 })}
      />

      <SliderInput
        label="Translation"
        min={0}
        max={1.5}
        step={0.05}
        selector={getTranslationSelectorFor(similarityId)}
        actionCreator={getTranslationSetterFor(similarityId)}
        valueFormatter={(value) => formatNumber(value, { precision: 2 })}
      />

      <SliderInput
        label="Depth"
        min={0}
        max={8}
        step={1}
        selector={getDepthSelectorFor(similarityId)}
        actionCreator={getDepthSetterFor(similarityId)}
        valueFormatter={(value) => formatNumber(value, { precision: 0 })}
      />
    </Inputs>
  </Container>
)

export default SimilarityControls
