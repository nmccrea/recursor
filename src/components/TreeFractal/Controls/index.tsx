import React from "react"
import { useSelector } from "react-redux"
import { selectIds } from "state/similarities/selectors"
import {
  Panel,
  GlobalControlsArea,
  SimilarityControlsArea,
  Instructions,
} from "./styled"
import GithubLink from "./GithubLink"
import CreateSimilarityButton from "./CreateSimilarityButton"
import SimilarityControls from "./SimilarityControls"

const Controls = () => {
  const similarityIds = useSelector(selectIds)
  return (
    <Panel>
      <GlobalControlsArea>
        <CreateSimilarityButton />
        <Instructions>{"< Add up to 3 similarities"}</Instructions>
        <GithubLink />
      </GlobalControlsArea>

      <SimilarityControlsArea>
        {similarityIds.map((similarityId) => (
          <SimilarityControls similarityId={similarityId} key={similarityId} />
        ))}
      </SimilarityControlsArea>
    </Panel>
  )
}

export default Controls
