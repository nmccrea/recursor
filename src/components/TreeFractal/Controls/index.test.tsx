import React from "react"
import { render } from "@testing-library/react"
import * as reactRedux from "react-redux"
import { RootState } from "../../../state/store"
import Controls from "."

describe("`<Controls />`", () => {
  it("renders correctly when there are no existing similarities", () => {
    const mockState: RootState = { similarities: { ids: [], entities: {} } }
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((selector) => selector(mockState))
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(<Controls />)

    expect(container).toMatchSnapshot()
  })

  it("renders correctly when there is one existing similarity", () => {
    const mockState: RootState = {
      similarities: {
        ids: ["test/subject"],
        entities: {
          "test/subject": {
            id: "test/subject",
            translation: 0.4663,
            scale: 0.7791,
            angle: 0.1385,
            depth: 3,
            color: "gray",
          },
        },
      },
    }
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((selector) => selector(mockState))

    const { container } = render(<Controls />)

    expect(container).toMatchSnapshot()
  })

  it("renders correctly when there are multiple existing similarities.", () => {
    const mockState: RootState = {
      similarities: {
        ids: ["test/a", "test/b"],
        entities: {
          "test/a": {
            id: "test/subject",
            translation: 0.4663,
            scale: 0.7791,
            angle: 0.1385,
            depth: 3,
            color: "gray",
          },
          "test/b": {
            id: "test/b",
            translation: 0.7432,
            scale: 0.7202,
            angle: 0.4692,
            depth: 7,
            color: "brown",
          },
        },
      },
    }
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((selector) => selector(mockState))

    const { container } = render(<Controls />)

    expect(container).toMatchSnapshot()
  })
})
