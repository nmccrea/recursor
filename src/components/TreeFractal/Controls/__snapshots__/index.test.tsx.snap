// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<Controls />\` renders correctly when there are multiple existing similarities. 1`] = `
.c0 {
  padding: 0.5rem;
  box-shadow: 0 -3px 5px 2px #bbb;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  overflow-y: auto;
}

.c1 {
  padding-bottom: 0.5rem;
}

.c3 {
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
}

.c2 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: cornflowerblue;
}

.c2:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c2:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

.c5 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: lightsalmon;
}

.c5:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c5:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-right: 1rem;
}

.c6 {
  padding: 0.5rem 0;
}

.c6 + div {
  border-top: 1px solid lightgray;
}

.c7 {
  display: grid;
  grid-template-columns: repeat(3,minmax(0,1fr));
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c8 {
  grid-column: span 3 / span 3;
  font-size: 0.9em;
}

.c9 {
  grid-column: span 2 / span 2;
}

.c10 {
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
}

.c12 {
  justify-self: flex-start;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c13 {
  justify-self: flex-end;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c11 {
  font-size: 0.8em;
  font-style: italic;
  color: gray;
}

@media (min-width:640px) {
  .c3 {
    grid-template-columns: repeat(2,minmax(0,1fr));
  }
}

@media (min-width:768px) {
  .c3 {
    grid-template-columns: repeat(3,minmax(0,1fr));
  }
}

@media (min-width:1024px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:1024px) {
  .c4 {
    padding: 0.5rem 0;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media (min-width:1024px) {
  .c6 {
    padding: 0 0.5rem;
  }

  .c6 + div {
    border-top: unset;
    border-left: 1px solid lightgray;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="sc-AxhCb c2"
      >
        Add
      </button>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div>
          <button
            class="sc-AxhCb c5"
          >
            Remove
          </button>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/a-angle"
            >
              Angle
            </label>
            <input
              class="c9"
              id="test/a-angle"
              max="360"
              min="-360"
              step="1"
              type="range"
              value="180"
            />
            <span
              class="c10"
            >
              <output
                for="test/a-angle"
              >
                180
              </output>
              <span
                class="c11"
              >
                 
                °
              </span>
            </span>
            <span
              class="c12"
            >
              -360
            </span>
            <span
              class="c13"
            >
              360
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/a-scale"
            >
              Scale
            </label>
            <input
              class="c9"
              id="test/a-scale"
              max="2"
              min="0"
              step="0.05"
              type="range"
              value="0.7791"
            />
            <span
              class="c10"
            >
              <output
                for="test/a-scale"
              >
                0.78
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              2.00
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/a-translation"
            >
              Translation
            </label>
            <input
              class="c9"
              id="test/a-translation"
              max="1.5"
              min="0"
              step="0.05"
              type="range"
              value="0.4663"
            />
            <span
              class="c10"
            >
              <output
                for="test/a-translation"
              >
                0.47
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              1.50
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/a-depth"
            >
              Depth
            </label>
            <input
              class="c9"
              id="test/a-depth"
              max="8"
              min="0"
              step="1"
              type="range"
              value="3"
            />
            <span
              class="c10"
            >
              <output
                for="test/a-depth"
              >
                3
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0
            </span>
            <span
              class="c13"
            >
              8
            </span>
          </div>
        </div>
      </div>
      <div
        class="c4"
      >
        <div>
          <button
            class="sc-AxhCb c5"
          >
            Remove
          </button>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/b-angle"
            >
              Angle
            </label>
            <input
              class="c9"
              id="test/b-angle"
              max="360"
              min="-360"
              step="1"
              type="range"
              value="-180"
            />
            <span
              class="c10"
            >
              <output
                for="test/b-angle"
              >
                -180
              </output>
              <span
                class="c11"
              >
                 
                °
              </span>
            </span>
            <span
              class="c12"
            >
              -360
            </span>
            <span
              class="c13"
            >
              360
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/b-scale"
            >
              Scale
            </label>
            <input
              class="c9"
              id="test/b-scale"
              max="2"
              min="0"
              step="0.05"
              type="range"
              value="0.7202"
            />
            <span
              class="c10"
            >
              <output
                for="test/b-scale"
              >
                0.72
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              2.00
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/b-translation"
            >
              Translation
            </label>
            <input
              class="c9"
              id="test/b-translation"
              max="1.5"
              min="0"
              step="0.05"
              type="range"
              value="0.7432"
            />
            <span
              class="c10"
            >
              <output
                for="test/b-translation"
              >
                0.74
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              1.50
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/b-depth"
            >
              Depth
            </label>
            <input
              class="c9"
              id="test/b-depth"
              max="8"
              min="0"
              step="1"
              type="range"
              value="7"
            />
            <span
              class="c10"
            >
              <output
                for="test/b-depth"
              >
                7
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0
            </span>
            <span
              class="c13"
            >
              8
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`\`<Controls />\` renders correctly when there are no existing similarities 1`] = `
.c0 {
  padding: 0.5rem;
  box-shadow: 0 -3px 5px 2px #bbb;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  overflow-y: auto;
}

.c1 {
  padding-bottom: 0.5rem;
}

.c3 {
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
}

.c2 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: cornflowerblue;
}

.c2:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c2:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

@media (min-width:640px) {
  .c3 {
    grid-template-columns: repeat(2,minmax(0,1fr));
  }
}

@media (min-width:768px) {
  .c3 {
    grid-template-columns: repeat(3,minmax(0,1fr));
  }
}

@media (min-width:1024px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="sc-AxhCb c2"
      >
        Add
      </button>
    </div>
    <div
      class="c3"
    />
  </div>
</div>
`;

exports[`\`<Controls />\` renders correctly when there is one existing similarity 1`] = `
.c0 {
  padding: 0.5rem;
  box-shadow: 0 -3px 5px 2px #bbb;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  overflow-y: auto;
}

.c1 {
  padding-bottom: 0.5rem;
}

.c3 {
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
}

.c2 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: cornflowerblue;
}

.c2:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c2:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

.c5 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: lightsalmon;
}

.c5:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c5:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-right: 1rem;
}

.c6 {
  padding: 0.5rem 0;
}

.c6 + div {
  border-top: 1px solid lightgray;
}

.c7 {
  display: grid;
  grid-template-columns: repeat(3,minmax(0,1fr));
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c8 {
  grid-column: span 3 / span 3;
  font-size: 0.9em;
}

.c9 {
  grid-column: span 2 / span 2;
}

.c10 {
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
}

.c12 {
  justify-self: flex-start;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c13 {
  justify-self: flex-end;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c11 {
  font-size: 0.8em;
  font-style: italic;
  color: gray;
}

@media (min-width:640px) {
  .c3 {
    grid-template-columns: repeat(2,minmax(0,1fr));
  }
}

@media (min-width:768px) {
  .c3 {
    grid-template-columns: repeat(3,minmax(0,1fr));
  }
}

@media (min-width:1024px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:1024px) {
  .c4 {
    padding: 0.5rem 0;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media (min-width:1024px) {
  .c6 {
    padding: 0 0.5rem;
  }

  .c6 + div {
    border-top: unset;
    border-left: 1px solid lightgray;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="sc-AxhCb c2"
      >
        Add
      </button>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div>
          <button
            class="sc-AxhCb c5"
          >
            Remove
          </button>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/subject-angle"
            >
              Angle
            </label>
            <input
              class="c9"
              id="test/subject-angle"
              max="360"
              min="-360"
              step="1"
              type="range"
              value="90"
            />
            <span
              class="c10"
            >
              <output
                for="test/subject-angle"
              >
                90
              </output>
              <span
                class="c11"
              >
                 
                °
              </span>
            </span>
            <span
              class="c12"
            >
              -360
            </span>
            <span
              class="c13"
            >
              360
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/subject-scale"
            >
              Scale
            </label>
            <input
              class="c9"
              id="test/subject-scale"
              max="2"
              min="0"
              step="0.05"
              type="range"
              value="0.7791"
            />
            <span
              class="c10"
            >
              <output
                for="test/subject-scale"
              >
                0.78
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              2.00
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/subject-translation"
            >
              Translation
            </label>
            <input
              class="c9"
              id="test/subject-translation"
              max="1.5"
              min="0"
              step="0.05"
              type="range"
              value="0.4663"
            />
            <span
              class="c10"
            >
              <output
                for="test/subject-translation"
              >
                0.47
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0.00
            </span>
            <span
              class="c13"
            >
              1.50
            </span>
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              class="c8"
              for="test/subject-depth"
            >
              Depth
            </label>
            <input
              class="c9"
              id="test/subject-depth"
              max="8"
              min="0"
              step="1"
              type="range"
              value="3"
            />
            <span
              class="c10"
            >
              <output
                for="test/subject-depth"
              >
                3
              </output>
              <span
                class="c11"
              >
                 
                x
              </span>
            </span>
            <span
              class="c12"
            >
              0
            </span>
            <span
              class="c13"
            >
              8
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
